### PR TITLE
Fix ICE in decimal_literal_representation lint

### LINKED ISSUE #3891: Internal compiler error due to invalid integer literal
```rust
fn main() {
    1x;
}
```

The invalid integer literal `1x` causes clippy to panic.

```
error: invalid suffix `x` for numeric literal
 --> src/main.rs:2:5
  |
2 |     1x;
  |     ^^ invalid suffix `x`
  |
  = help: the suffix must be one of the integral types (`u32`, `isize`, etc)

thread 'rustc' panicked at 'called `Result::unwrap()` on an `Err` value: ParseIntError { kind: InvalidDigit }', src/libcore/result.rs:997:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:215
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:482
   6: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:385
   7: rust_begin_unwind
             at src/libstd/panicking.rs:312
   8: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   9: core::result::unwrap_failed
  10: <clippy_lints::literal_representation::LiteralRepresentation as rustc::lint::EarlyLintPass>::check_expr
  11: <rustc::lint::context::EarlyContext<'a> as syntax::visit::Visitor<'a>>::visit_expr
  12: <rustc::lint::context::EarlyContext<'a> as syntax::visit::Visitor<'a>>::visit_block
  13: <rustc::lint::context::EarlyContext<'a> as syntax::visit::Visitor<'a>>::visit_fn
  14: syntax::visit::walk_item
  15: <rustc::lint::context::EarlyContext<'a> as syntax::visit::Visitor<'a>>::visit_item
  16: <rustc::lint::context::EarlyContext<'a> as syntax::visit::Visitor<'a>>::visit_mod
  17: rustc::lint::context::check_ast_crate
  18: rustc::util::common::time
  19: rustc_driver::driver::phase_2_configure_and_expand
  20: rustc_driver::driver::compile_input
  21: rustc_driver::run_compiler_with_pool
  22: <scoped_tls::ScopedKey<T>>::set
  23: rustc_driver::run_compiler
  24: <scoped_tls::ScopedKey<T>>::set
query stack during panic:
end of query stack
error: aborting due to previous error


error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.33.0 (2aa4c46cf 2019-02-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden
```

### PR DESCRIPTION
**Pull request by phansch**
Handling the integer parsing properly instead of just unwrapping.

Note that the test is not catching the ICE because plain UI tests
[currently hide ICEs][compiletest_issue]. Once that issue is fixed, this
test would fail properly again.

Fixes #3891

[compiletest_issue]: https://github.com/laumann/compiletest-rs/issues/169

### CODE REVIEW

**Comment by phansch** [comment]
For some reason rustfmt is triggering the same rust warning: https://travis-ci.com/rust-lang/rust-clippy/jobs/191281098#L1664
I didn't know that running rustfmt will trigger rustc errors as well and I'm not sure how to properly test this then :confused:

**Comment by flip1995** [comment]
Maybe use the old hack and exclude the test file here:
https://github.com/rust-lang/rust-clippy/blob/949f58440b19e470a8ed370166f9d72cd09bd7e3/ci/base-tests.sh#L62

`rustfmt` only triggers parsing errors, since the file needs to be parsed for `rustfmt` to run.

**Comment by phansch** [comment]
I re-added the old hack to exclude the file from rustfmt. I think in this specific case it makes sense to be excluded.

File: clippy_lints/src/literal_representation.rs [hunk 1]
```diff
@@ -529,19 +529,23 @@ impl LiteralRepresentation {
...
+                            warning_type.display(&digit_info.grouping_hint(), cx, lit.span)
+                        });
                     }
-                    let hex = format!("{:#X}", val);
-                    let digit_info = DigitInfo::new(&hex[..], false);
-                    let _ = Self::do_lint(digit_info.digits).map_err(|warning_type| {
-                        warning_type.display(&digit_info.grouping_hint(), cx, lit.span)
-                    });
+                    else {
+                        return
```
Comment (flip1995) [inline clippy_lints/src/literal_representation.rs]: NIT: `else` formatting

Is this `else { return }` required? It's not in a loop and the function ends after the `if_chain!` macro.

File: clippy_lints/src/literal_representation.rs [hunk 2]
```diff
@@ -529,19 +529,23 @@ impl LiteralRepresentation {
...
                         .collect::<String>()
-                        .parse::<u128>().unwrap();
-                    if val < u128::from(self.threshold) {
-                        return
+                        .parse::<u128>() {
+                        if val < u128::from(self.threshold) {
+                            return
+                        }
+                        let hex = format!("{:#X}", val);
+                        let digit_info = DigitInfo::new(&hex[..], false);
```
Comment (flip1995) [inline clippy_lints/src/literal_representation.rs]:
```suggestion
                        let digit_info = DigitInfo::new(&hex, false);
```
should also work, shouldn't it?

File: clippy_lints/src/literal_representation.rs (hunk 1 above)
Comment (phansch) [reply clippy_lints/src/literal_representation.rs]: Good catch, no it's not required =)

**Comment by flip1995** [comment]
One last thing: All these `if` statements should be collapsible into the `if_chain!` macro:

```rust
if_chain! {
    if let LitKind::Int(..) = lit.node;
    if let Some(src) = snippet_opt(cx, lit.span);
    if let Some(firstch) = src.chars().next();
    if char::to_digit(firstch, 10).is_some();
    let digit_info = DigitInfo::new(&src, false);
    if digit_info.radix == Radix::Decimal;
    if let Ok(val) = digit_info.digits;
    if val >= u128::from(self.threshold);
    then {
        let hex = format!("{:#X}", val);
        let digit_info = DigitInfo::new(&hex, false);
        let _ = Self::do_lint(digit_info.digits).map_err(|warning_type| {
            warning_type.display(&digit_info.grouping_hint(), cx, lit.span)
        });
    }
}
```

**Comment by phansch** [comment]
Yup, that worked =)

**Pull request merged**